[heft] using both heft plugins for webpack4 and webpack5 in a project can cause webpack type errors

<!-- Have a question?  Before creating an issue, ask in the chat room: https://rushstack.zulipchat.com/ -->

<!-- Invoke Heft with the "--debug" parameter.  It often prints information that is helpful for diagnosing problems. -->

<!--------------------------------------------------------------------------
👉 STEP 1: Write a concise but specific issue title in the box above.
     Example: "[heft] Linting property getter causes TypeError"
--------------------------------------------------------------------------->

## Summary
typescript will throw errors when you have both webpack 5 and webpack 4 rigs in a project 
<!--------------------------------------------------------------------------
👉 STEP 2: In a few sentences, please explain:

     What were you trying to accomplish?
     What action did you perform that ran into trouble?
     What went wrong?
--------------------------------------------------------------------------->

## Repro steps
##[error][typescript] Error: ../../../common/temp/node_modules/.pnpm/@rushstack+module-minifier-plugin@0.9.2_8a934206096b177ef4e8c472548f577c/node_modules/@rushstack/module-minifier-plugin/dist/module-minifier-plugin.d.ts:17:10 - (TS2305) Module '"webpack"' has no exported member 'Plugin'.
##[error][typescript] Error: ../../../common/temp/node_modules/.pnpm/@rushstack+module-minifier-plugin@0.9.2_8a934206096b177ef4e8c472548f577c/node_modules/@rushstack/module-minifier-plugin/dist/module-minifier-plugin.d.ts:39:74 - (TS2724) 'exports' has no exported member named 'compilation'. Did you mean 'Compilation'?
<!--------------------------------------------------------------------------
👉 STEP 3: If your issue is a feature request and not a bug, delete this
     "Repro steps" section and skip to STEP 6.

👉 STEP 4: In many cases we can investigate bugs much faster if you include:
     The URL for a simplified Git branch that reproduces the problem.
     Step by step instructions for how to build the branch and see the error.

👉 STEP 5: It's also helpful to include an "expected" and "actual" result.
     But if that's not relevant, feel free to delete those fields.
--------------------------------------------------------------------------->

  **Expected result:** <!-- What you expected these steps to accomplish -->

  **Actual result:** <!-- If an error occurred, include the full error message text and any call stack. -->

## Details
Adding @types/webpack as a peer dependency in heft-webpack4-plugin should solve this problem

--- a/common/changes/@rushstack/heft-webpack4-plugin/selarkin-issue-3463_2022-06-13-19-32.json
+++ b/common/changes/@rushstack/heft-webpack4-plugin/selarkin-issue-3463_2022-06-13-19-32.json
@@ -1,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@rushstack/heft-webpack4-plugin",
+      "comment": "add optional peer depedency on webpack types",
+      "type": "patch"
+    }
+  ],
+  "packageName": "@rushstack/heft-webpack4-plugin"
+}

--- a/heft-plugins/heft-webpack4-plugin/package.json
+++ b/heft-plugins/heft-webpack4-plugin/package.json
@@ -19,6 +19,7 @@
   },
   "peerDependencies": {
     "@rushstack/heft": "^0.45.6",
+    "@types/webpack": "4.41.24",
     "webpack": "~4.44.2"
   },
   "dependencies": {

--- a/heft-plugins/heft-webpack4-plugin/package.json
+++ b/heft-plugins/heft-webpack4-plugin/package.json
@@ -17,9 +17,14 @@
     "_phase:build": "heft build --clean",
     "_phase:test": "heft test --no-build"
   },
+  "peerDependenciesMeta": {
+    "@types/webpack": {
+      "optional": true
+    }
+  },
   "peerDependencies": {
     "@rushstack/heft": "^0.45.6",
-    "@types/webpack": "4.41.24",
+    "@types/webpack": "^4",
     "webpack": "~4.44.2"
   },
   "dependencies": {

